Master wowl mrp documents kanban wbr

Williambraecky opened the pull request:
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/99769).

Williambraecky commented:
@odoo/rd-discuss I took the liberty of removing `DocumentViewer` as it was not used anymore. This view was more or less a simpler version of the documents kanban, reusing a lot of the same stuff (even the tests are very similar)

@odoo/rd-framework-js Same thing for `file_upload_mixin` (and its related files) I removed it completely, replaced by the `file_upload` service and hooks, though I did not put it in the `web` module, if you agree to move it in the `web` module I can reuse this code in documents as well and reduce duplicated code.
Though I did not write explicit code for the service, the hooks are tested implicitely.

@odoo/rd-logistics If you could test it as well it would be great :)
Thanks

seb-odoo reviewed (approved):
Nice to see the removal of old `DocumentViewer`, you could as well look for related classes in css files and remove them too if no longer used elsewhere.

FrancoisGe commented on addons/mrp/static/src/views/mrp_documents_kanban/helper/file_upload_service.js:
```diff
@@ -0,0 +1,107 @@
+    },
+
+    _uploads: reactive({}),
+    _nextId: 1,
```
Why not define it in start?

FrancoisGe commented on addons/mrp/static/tests/mrp_document_kanban_tests.js:
```diff
@@ -23,53 +34,55 @@ QUnit.module('MrpDocumentsKanbanView', {
-        MrpDocumentsKanbanController.prototype._createXHR = this.ORIGINAL_CREATE_XHR;
+        fileUploadService.createXhr = this.ORIGINAL_CREATE_XHR;
+        // Clear the uploads as we don't want to keep them between tests
+        serviceRegistry.get("file_upload")._uploads = reactive({});
```
Not necessary if defined in the start.

FrancoisGe commented on addons/mrp/static/src/views/mrp_documents_kanban/helper/file_upload_hook.js:
```diff
@@ -0,0 +1,61 @@
+    const mainComponentCategory = registry.category("main_components");
+    if (mainComponentCategory.contains(FileUploadProgressContainerKey)) {
+        mainComponentCategory.remove(FileUploadProgressContainerKey);
+    }
```
I'm not sure I understand this part ? And I don't know if this is a good practice.

Williambraecky commented on addons/mrp/static/src/views/mrp_documents_kanban/mrp_documents_kanban_record.js:
```diff
@@ -0,0 +1,36 @@
+            window.location = `/web/content/mrp.document/${this.props.record.resId}/datas?download=true`;
+            return;
+        } else if (ev.target.closest(".o_kanban_previewer")) {
+            this.env.services.messaging.get().then((messaging) => {
```
The function isn't async and it's an override of a sync function, is it safe to change the sync/async stuff through an override?

ged-odoo replied:
if the function is not async, then it should not be called with `then` or async/await

ged-odoo commented on addons/web/static/src/core/file_upload/file_upload_progress_bar.js:
```diff
@@ -0,0 +1,28 @@
+
+export class FileUploadProgressBar extends Component {
+    setup() {
+        this.dialogService = this.env.services.dialog;
```
never get the service like this, always use `useService` instead

ged-odoo commented on addons/web/static/src/core/file_upload/file_upload_progress_bar.js:
```diff
@@ -0,0 +1,28 @@
+/** @odoo-module **/
+
+import { sprintf } from "@web/core/utils/strings";
```
you should probably use relative imports inside `core`

ged-odoo commented on addons/web/static/src/core/file_upload/file_upload_progress_bar.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates>
+    <t t-name="web.FileUploadProgressBar" owl="1">
+        <div class="o_file_upload_progress_bar">
```
do you need all these custom classes? can't you do it with just plain bootstrap? also, if you do need to use these classes,  they should be defined in a css file next to this template then

ged-odoo commented on addons/web/static/src/core/file_upload/file_upload_progress_record.js:
```diff
@@ -0,0 +1,35 @@
+/** @odoo-module **/
+
+import { sprintf } from "@web/core/utils/strings";
+import { FileUploadProgressBar } from "@web/core/file_upload/file_upload_progress_bar";
```
relative imports please!

sdegueldre commented on addons/web/static/src/core/file_upload/file_upload_service.js:
```diff
@@ -0,0 +1,106 @@
+            return upload;
+        };
+
+        const getUpload = (id) => _uploads[id];
+        const getUploads = () => _uploads;
```
these functions are redundant with one another, the second function also exposes the _uploads outright, at which point a getter is pretty useless, just put the uploads object directly in the returned object.

ged-odoo commented on addons/web/static/src/core/file_upload/file_upload_progress_record.xml:
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<templates>
+    <t t-name="web.FileUploadProgressKanbanRecord" owl="1">
```
not sure why do need to reference kanban view here. it would be better if this component was completely independant

Williambraecky replied:
Keep in mind that this change is meant to replace what was previously available, maybe the file should be moved closer to `kanban` instead?

ged-odoo commented on addons/web/static/src/core/file_upload/file_upload_service.js:
```diff
@@ -0,0 +1,106 @@
+         */
+        const upload = async (route, files, params = {}) => {
+            const xhr = this.createXhr();
+            const bus = params.bus || env.bus;
```
do you really need a bus here? please do not use the env.bus. your service should export a bus instead

ged-odoo commented on addons/web/static/src/core/file_upload/file_upload_hook.js:
```diff
@@ -0,0 +1,18 @@
+const { useComponent } = owl;
+
+export function useFileUpload({ onAdded, onLoaded, onError }) {
+    const component = useComponent();
```
this should simply import the bus directly from the service: 
```js
const { bus } = useService("file_upload");
```

ged-odoo reviewed (commented):
mostly ok in my opinion. however, I do not see any tests in web... kind of scary. would be better to have a few tests (I know that the file_upload service is itself difficult to test, but the components themselves could be tested with a mock file_upload service)

Williambraecky commented:
> mostly ok in my opinion. however, I do not see any tests in web... kind of scary. would be better to have a few tests (I know that the file_upload service is itself difficult to test, but the components themselves could be tested with a mock file_upload service)

I have no idea how to test the file upload service..
It is tested in the mrp documents kanban tests, which can be partially moved to `web` :)

Thanks a lot for reviewing.

Williambraecky commented on addons/web/static/src/core/file_upload/file_upload_service.js:
```diff
@@ -0,0 +1,106 @@
+                delete _uploads[upload.id];
+                upload.state = "error";
+                // Disable this option if you need more explicit error handling.
+                if (params.displayErrorNotification !== undefined && params.displayErrorNotification) {
```
It's to make the option non required

Williambraecky commented:
@ged-odoo would be awesome if you could check again :)

ged-odoo reviewed (approved):
looks good enough to me

tivisse commented:
@robodoo r+
